rascal-lsp/src/main/rascal/demo/lang/pico/LanguageServer.rsc

```diff
@@ -44,30 +44,43 @@ set[LanguageService] picoLanguageContributor() = {
 
 set[LanguageService] picoLanguageContributorSlowSummary() = {
     parser(parser(#start[Program])),
-    summarizer(picoSummarizer, providesImplementations = false)
+    analyzer(picoAnalyzer, providesImplementations = false),
+    builder(picoBuilder)
 };
 
 list[DocumentSymbol] picoOutliner(start[Program] input)
   = [symbol("<input.src>", DocumentSymbolKind::\file(), input.src, children=[
       *[symbol("<var.id>", \variable(), var.src) | /IdType var := input]
   ])];
 
-Summary picoSummarizer(loc l, start[Program] input) {
-    println("Running summary for pico!");
-    rel[str, loc] defs = {<"<var.id>", var.src> | /IdType var  := input};
+Summary picoAnalyzer(loc l, start[Program] input) {
+    println("Running analyzer for pico!");
+    rel[str, loc] defs = {<"<var.id>", var.src> | /IdType var := input};
     rel[loc, str] uses = {<id.src, "<id>"> | /Id id := input};
+    rel[loc, str] asgn = {<id.src, "<id>"> | /Statement stmt := input, (Statement) `<Id id> := <Expression _>` := stmt};
     rel[loc, str] docs = {<var.src, "*variable* <var>"> | /IdType var := input};
 
-
     return summary(l,
-        messages = {<src, error("<id> is not defined", src)> | <src, id> <- uses, id notin defs<0>},
+        messages = {<src, warning("<id> is not assigned", src)> | <id, src> <- defs, id notin asgn<1>},
         references = (uses o defs)<1,0>,
         definitions = uses o defs,
         documentation = (uses o defs) o docs
     );
 }
 
-set[loc] lookupDef(loc l, start[Program] input, Tree cursor) =
+Summary picoBuilder(loc l, start[Program] input) {
```

Comment:
for the case of the example, what do you think of the builder calling the analyzer, and then adding some more checks? (or better: sharing a part of the light weight typechecker/namebinding part)